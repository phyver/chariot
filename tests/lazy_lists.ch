:set show_priorities false

(* type of natural numbers *)
data nat where Zero : nat | Succ : nat -> nat

(* addition of natural numbers *)
val add : nat -> nat -> nat
    | add n 0 = n
    | add n (m+1) = Succ (add n m)

val mult : nat -> nat -> nat
    | mult n 0 = 0
    | mult n (m+1) = n + (mult n m)

val pow : nat -> nat -> nat
    | pow n 0 = 1
    | pow n (m+1) = mult n (pow n m)


data list_aux('a,'l) where Nil : list_aux('a,'l) | Cons : 'a -> 'l -> list_aux('a,'l)

codata colist('a) where Unfreeze : colist('a) -> list_aux('a,colist('a))

val freeze : list_aux('a,colist('a)) -> colist('a)
  | (freeze x).Unfreeze = x

val
    (map f s).Unfreeze = map_aux f s.Unfreeze
and
    | map_aux f (Nil) = Nil
    | map_aux f (Cons x s) = Cons (f x) (map f s)

val zeros.Unfreeze = Cons 0 zeros

val (incr n).Unfreeze = Cons n (incr (n+1))

val   (decr 0).Unfreeze = Nil
    | (decr (n+1)).Unfreeze = Cons n (decr n)

:reduce incr 10

:reduce decr 10

:unfold incr 10, 5

:unfold incr 10, 12

:unfold decr 10, 5

:unfold decr 10, 10

:unfold decr 10, 20










