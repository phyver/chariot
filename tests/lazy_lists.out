……… term: incr 10
……… result: incr 10
……… of type: colist(nat)

……… term: decr 10
……… result: decr 10
……… of type: colist(nat)

……… term: incr 10
……… result (at depth 5): {Unfreeze=10::{Unfreeze=11::{Unfreeze=12::{Unfreeze=13::{Unfreeze=14::{…<1>…}}}}}}
……… of type: colist(nat)

……… term: incr 10
……… result (at depth 12): {Unfreeze=10::{Unfreeze=11::{Unfreeze=12::{Unfreeze=13::{Unfreeze=14::{Unfreeze=15::{Unfreeze=16::{Unfreeze=17::{Unfreeze=18::{Unfreeze=19::{Unfreeze=20::{Unfreeze=21::{…<1>…}}}}}}}}}}}}}
……… of type: colist(nat)

……… term: decr 10
……… result (at depth 5): {Unfreeze=9::{Unfreeze=8::{Unfreeze=7::{Unfreeze=6::{Unfreeze=5::{…<1>…}}}}}}
……… of type: colist(nat)

……… term: decr 10
……… result (at depth 10): {Unfreeze=9::{Unfreeze=8::{Unfreeze=7::{Unfreeze=6::{Unfreeze=5::{Unfreeze=4::{Unfreeze=3::{Unfreeze=2::{Unfreeze=1::{Unfreeze=0::{…<1>…}}}}}}}}}}}
……… of type: colist(nat)

……… term: decr 10
……… result (at depth 20): {Unfreeze=9::{Unfreeze=8::{Unfreeze=7::{Unfreeze=6::{Unfreeze=5::{Unfreeze=4::{Unfreeze=3::{Unfreeze=2::{Unfreeze=1::{Unfreeze=0::{Unfreeze=[]}}}}}}}}}}}
……… of type: colist(nat)

